crates/bevy_ptr/src/lib.rs

```diff
@@ -7,17 +7,37 @@ use core::{
     cell::UnsafeCell, marker::PhantomData, mem::ManuallyDrop, num::NonZeroUsize, ptr::NonNull,
 };
 
+#[derive(Copy, Clone)]
+/// Used as a type argument to [`Ptr`], [`PtrMut`] and [`OwningPtr`] to specify that the pointer is aligned.
+pub struct Aligned;
+#[derive(Copy, Clone)]
+/// Used as a type argument to [`Ptr`], [`PtrMut`] and [`OwningPtr`] to specify that the pointer is not aligned.
+pub struct Unaligned;
+
+/// Trait that is only implemented for [`Aligned`] and [`Unaligned`] to work around the lack of ability
+/// to have const generics of an enum.
+pub trait IsAligned: sealed::Sealed {}
+impl IsAligned for Aligned {}
```

Comment:
Actually wonder if we could use const generics to enforce debug checks on `byte_add` and `byte_offset`, but I'll leave that for a separate PR.

Reply:
you wouldnt need const generics, it ought to be possible to look at the `TypeId` of `A` to see if its `Aligned` or not